ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java

```diff
@@ -873,6 +883,34 @@ protected void maybeProcessCallbacks() {
       }
     }
 
+    /**
+     * Decompress the compressed buffer specified.
+     * If the buffer is not compressed, based on blobCompressed property, then return the input buffer.
+     *
+     * @param sourceBuffer The source buffer to decompress.
+     * @return New buffer if decompressed successfully and the source buffer is released.
+     *         Original buffer is blob not compressed.  NULL if failed and source buffer is released.
+     */
+    protected ByteBuf decompressContent(ByteBuf sourceBuffer) {
+      if (!isChunkCompressed) {
+        // Blob is not compressed.  Return source buffer.
+        return sourceBuffer;
+      }
+
+      // Note: decompress() already emit metrics to for decompress() failure.
+      try {
+        ByteBuf result = decompressionService.decompress(sourceBuffer, routerConfig.routerMaxPutChunkSizeBytes);
+        sourceBuffer.release();
+        return result;
+      } catch (Exception ex) {
+        logger.error("Failed to decompress chunk data for chunkBlobId " + chunkBlobId + " due to exception.", ex);
+        setOperationException(buildChunkException("Exception thrown on decompressing the data chunk", ex,
+            RouterErrorCode.UnexpectedInternalError));
+        sourceBuffer.release();
+        return null;
+      }
```

Comment:
Handling decompression exceptions needs more thoughts. We are not going to serve data back to client when the decompression fails, so we have to set an exception for the entire operation. We can also set the operationCompleted to true so the operation would be stopped right away.

Reply:
Good idea.  Thank you.